stargz/fs.go

```diff
@@ -157,7 +162,100 @@ type connection struct {
 	lastCheck time.Time
 }
 
+func newPreemptiveTaskManager(concurrency int64, interval time.Duration) *preemptiveTaskManager {
+	return &preemptiveTaskManager{
+		preemptiveSem:          semaphore.NewWeighted(concurrency),
+		privilegedTaskInterval: interval,
+	}
+}
+
+// preemptiveTaskManager manages privileged and preemptive task execution.
+// Preemptive tasks are some tasks aren't important and we don't want them
+// to use compute resources(CPU, NW, etc...) during more important tasks(=
+// privileged tasks) running.
+type preemptiveTaskManager struct {
+	privilegedTasks        int64
+	preemptiveSem          *semaphore.Weighted
+	privilegedTaskInterval time.Duration
+}
+
+// doPrivilegedTask tells the manager that we are running a privileged task.
+func (ts *preemptiveTaskManager) doPrivilegedTask() {
+	atomic.AddInt64(&ts.privilegedTasks, 1)
+}
+
+// doPrivilegedTask tells the manager that we've done a privileged task.
+func (ts *preemptiveTaskManager) donePrivilegedTask() {
+	atomic.AddInt64(&ts.privilegedTasks, -1)
+}
+
+// invokePreemptiveTask invokes a preemptive task. The task is started only when
+// no privileged tasks are running. Privileged task's execution stops the
+// execution of all preemptive tasks. Priemptive task must be immutable and
+// resumable because this function suddenly stop and restart it.
+func (ts *preemptiveTaskManager) invokePreemptiveTask(do func(context.Context), timeout time.Duration) {
+	privilegedTaskAdded := func() <-chan int64 {
+		ch := make(chan int64)
+		go func() {
+			for ts.privilegedTasks <= 0 {
+				time.Sleep(10 * time.Millisecond)
+			}
+			ch <- ts.privilegedTasks
+		}()
+		return ch
+	}
+
+try:
+	// if nobody add privileged tasks in specified interval, this is a
+	// chance for us to do some preemptive tasks
+	select {
+	case <-privilegedTaskAdded():
+		time.Sleep(time.Second)
```

Comment:
ditto

Reply:
We need timeout here so we use channel instead of cond. There aren't easy ways to use cond with timeout?